### PR TITLE
Update QoS Cube Selection to select an unreliable flow (if there is one) if the app doesn't care.

### LINKED ISSUE #527: Selection of QoS cube is not working correctly (all configurations) by setting the maxAllowableGap in FlowSpecification
use any configuration file that has a normal DIF between two hosts

I have the following qos cubes:

"qosCubes": [
                {
                    "name": "reliable",
                    "id": 1,
                    "partialDelivery": false,
                    "orderedDelivery": true
                },
                {
                    "name": "unreliable",
                    "id": 2,
                    "partialDelivery": true,
                    "orderedDelivery": false
                }
            ],

after enrollment run rina-echo-time with -g >= 0 (sets maxAllowableGap)

(rina-echo-time without -g option shown, which works , this sets maxAllowableGap to -1)
```
dstaesse@n:/usr/local/irati/bin$ ./rina-echo-time -c 10
28822(1427818362)#concurrency (DBG): Lockable created successfully
28822(1427818362)#concurrency (DBG): Lockable created successfully
28822(1427818362)#logs (DBG): New log level: INFO
28822(1427818362)#netlink-manager (INFO): Netlink socket connected to local port 28822 
Flow allocation time = 18ms
SDU size = 20, seq = 0, RTT = 254us
SDU size = 20, seq = 1, RTT = 239us

dstaesse@n:/usr/local/irati/bin$ ./rina-echo-time -c 10 -g 0
28849(1427818367)#concurrency (DBG): Lockable created successfully
28849(1427818367)#concurrency (DBG): Lockable created successfully
28849(1427818367)#logs (DBG): New log level: INFO
28849(1427818367)#netlink-manager (INFO): Netlink socket connected to local port 28849 
28849(1427818368)#rina-echo-time (ERR): Failed to allocate a flow
```

error log from IPCP will show " Could not find a QoS Cube"

```
20332(1427816743)#flow-allocator (DBG): Created flow allocator instance to manage the flow identified by portId 5 
20332(1427816743)#flow-allocator (ERR): Problems allocating flow: Could not find a QoS Cube
20332(1427816743)#timer (DBG): Waiting for the timer 7959264 to join
20332(1427816743)#timer (DBG): Timer with ID 7959264 ended
20332(1427816743)#timer (DBG): Waiting for the timer 7954576 to join
20332(1427816744)#timer (DBG): Timer with ID 7954576 ended
20332(1427816744)#syscalls (DBG): Gonna call syscall SYS_deallocatePortId (322)
20332(1427816744)#syscalls (ERR): Syscall deallocate port id failed: -1
20332(1427816744)#flow-allocator (ERR): Problems releasing port-id 5: Problems requesting the allocation/deallocation of a port-id
20332(1427816744)#netlink-manager (DBG): NL msg TX. Fam: 22; Opcode: 13_IPCM_ALLOC_FLOW_REQ_RES; Sport: 20332; Dport: 1; Seqnum: 5; Response; SIPCP: 2; DIPCP: 0
```
reason lies here (rinad/src/ipcp/default-flow-allocator-ps.cc)

```cpp
rina::QoSCube * FlowAllocatorPs::selectQoSCube(
                const rina::FlowSpecification& flowSpec)
{
	std::list<rina::QoSCube*> qosCubes = dm->getQoSCubes();
	if (flowSpec.maxAllowableGap < 0) {
		return *(qosCubes.begin());
	}

	std::list<rina::QoSCube*>::const_iterator iterator;
	rina::QoSCube* cube;
	for (iterator = qosCubes.begin(); iterator != qosCubes.end(); ++iterator) {
		cube = *iterator;
		if (cube->get_efcp_policies().is_dtcp_present()) {
			if (flowSpec.maxAllowableGap >= 0
					&& cube->get_efcp_policies().get_dtcp_configuration().is_rtx_control()) {
				return cube;
			}
		}
	}

	throw rina::Exception("Could not find a QoS Cube");
}
```
configuration has a "reliable" flow, so one of
```cpp
cube->get_efcp_policies().is_dtcp_present()
cube->get_efcp_policies().get_dtcp_configuration().is_rtx_control()
```
is probably not finding the correct information (could be the reliable qos cube not registering).

### PR DESCRIPTION
**Pull request by dstaesse**
fixes #527

setting maxAllowableGap to -1 will select a QoS cube with following preference:
unreliable with flow control
unreliable without flow control
first qos cube available

### CODE REVIEW

**Comment by dstaesse** [comment]
Hi Marc, that was vwall-specific, and I inadvertently merged that. I undid those changes.

**Comment by dstaesse** [comment]
@msune saw you noticed already but commented in the other post
Just tested the code, confirm it works as advertised, both using rina-echo-time and rina-tgen:

```
dstaesse@n:/usr/local/irati/bin$ cat ../var/log/test1.IRATI-1.log | grep qos
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes, class qoscube set, instance 10 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/unreliablewithflowcontrol, class qoscube, instance 13 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/reliablewithflowcontrol, class qoscube, instance 14 added to the RIB
3985(1428675144)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675219)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675223)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
dstaesse@n:/usr/local/irati/bin$ sudo echo "" > ../var/log/test1.IRATI-1.log 
-bash: ../var/log/test1.IRATI-1.log: Permission denied
dstaesse@n:/usr/local/irati/bin$ sudo 'echo "" > ../var/log/test1.IRATI-1.log' 
sudo: echo "" > ../var/log/test1.IRATI-1.log: command not found
dstaesse@n:/usr/local/irati/bin$ ./rina-echo-time -g 1
5168(1428675293)#concurrency (DBG): Lockable created successfully
5168(1428675293)#concurrency (DBG): Lockable created successfully
5168(1428675293)#logs (DBG): New log level: INFO
5168(1428675293)#netlink-manager (INFO): Netlink socket connected to local port 5168 
Flow allocation time = 4354us
SDU size = 20, seq = 0, RTT = 243us
dstaesse@n:/usr/local/irati/bin$ cat ../var/log/test1.IRATI-1.log | grep qos 
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes, class qoscube set, instance 10 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/unreliablewithflowcontrol, class qoscube, instance 13 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/reliablewithflowcontrol, class qoscube, instance 14 added to the RIB
3985(1428675144)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675219)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675223)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675293)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
dstaesse@n:/usr/local/irati/bin$ ./rina-echo-time -g -1
5215(1428675301)#concurrency (DBG): Lockable created successfully
5215(1428675301)#concurrency (DBG): Lockable created successfully
5215(1428675301)#logs (DBG): New log level: INFO
5215(1428675301)#netlink-manager (INFO): Netlink socket connected to local port 5215 
Flow allocation time = 4161us
SDU size = 20, seq = 0, RTT = 252us
dstaesse@n:/usr/local/irati/bin$ cat ../var/log/test1.IRATI-1.log | grep qos
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes, class qoscube set, instance 10 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/unreliablewithflowcontrol, class qoscube, instance 13 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/reliablewithflowcontrol, class qoscube, instance 14 added to the RIB
3985(1428675144)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675219)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675223)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675293)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675301)#flow-allocator-ps-default (DBG): Selected qos cube with name unreliablewithflowcontrol
dstaesse@n:/usr/local/irati/bin$ ./rina-tgen --duration 1 --rate 100 --qoscube unreliable
5992(1428675454)#concurrency (DBG): Lockable created successfully
5992(1428675454)#concurrency (DBG): Lockable created successfully
5992(1428675454)#logs (DBG): New log level: INFO
5992(1428675454)#netlink-manager (INFO): Netlink socket connected to local port 5992 
5992(1428675454)#traffic-generator (INFO): starting test
5992(1428675455)#traffic-generator (INFO): sent statistics: 26 SDUs, 13000 bytes in 1000000 us
5992(1428675455)#traffic-generator (INFO):      => 0.1040 Mb/s
^C
dstaesse@n:/usr/local/irati/bin$ cat ../var/log/test1.IRATI-1.log | grep qos
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes, class qoscube set, instance 10 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/unreliablewithflowcontrol, class qoscube, instance 13 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/reliablewithflowcontrol, class qoscube, instance 14 added to the RIB
3985(1428675144)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675219)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675223)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675293)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675301)#flow-allocator-ps-default (DBG): Selected qos cube with name unreliablewithflowcontrol
3985(1428675454)#flow-allocator-ps-default (DBG): Selected qos cube with name unreliablewithflowcontrol
dstaesse@n:/usr/local/irati/bin$ ./rina-tgen --duration 1 --rate 100 --qoscube reliable
6094(1428675473)#concurrency (DBG): Lockable created successfully
6094(1428675473)#concurrency (DBG): Lockable created successfully
6094(1428675473)#logs (DBG): New log level: INFO
6094(1428675473)#netlink-manager (INFO): Netlink socket connected to local port 6094 
6094(1428675473)#traffic-generator (INFO): starting test
6094(1428675474)#traffic-generator (INFO): sent statistics: 26 SDUs, 13000 bytes in 1000000 us
6094(1428675474)#traffic-generator (INFO):      => 0.1040 Mb/s
^C
dstaesse@n:/usr/local/irati/bin$ cat ../var/log/test1.IRATI-1.log | grep qos
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes, class qoscube set, instance 10 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/unreliablewithflowcontrol, class qoscube, instance 13 added to the RIB
3985(1428675066)#rib (INFO): Object with name /dif/management/flowallocator/qoscubes/reliablewithflowcontrol, class qoscube, instance 14 added to the RIB
3985(1428675144)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675219)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675223)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675293)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
3985(1428675301)#flow-allocator-ps-default (DBG): Selected qos cube with name unreliablewithflowcontrol
3985(1428675454)#flow-allocator-ps-default (DBG): Selected qos cube with name unreliablewithflowcontrol
3985(1428675473)#flow-allocator-ps-default (DBG): Selected qos cube with name reliablewithflowcontrol
dstaesse@n:/usr/local/irati/bin$ 
```

File: rinad/src/ipcp/default-flow-allocator-ps.cc [hunk 1]
```diff
@@ -99,24 +99,37 @@ Flow * FlowAllocatorPs::newFlowRequest(IPCProcess * ipc_process,
...
+        rina::QoSCube* cube;
+
+	if (*(qosCubes.begin())==NULL)
+	    throw rina::Exception("No QoSCubes defined.");
+
 	if (flowSpec.maxAllowableGap < 0) {
+	        for (iterator = qosCubes.begin(); iterator != qosCubes.end(); ++iterator) {
+		        cube = *iterator;
+		        if (cube->get_efcp_policies().is_dtcp_present()
+			    && !cube->get_efcp_policies().get_dtcp_configuration().is_rtx_control())				return cube;
```
Comment (vmaffione) [inline rinad/src/ipcp/default-flow-allocator-ps.cc]: "return cube" is not visible!

**Comment by vmaffione** [comment]
From the logic I see that if max_allowed_gap is -1 (+infinite), a QoSCube with DTCP (but without rtx) is preferred over a QoSCube without DTCP. Is that what we want? If yes, what is the reasoning behind this choice?

**Comment by dstaesse** [comment]
there is no other way to select without DTCP otherwise. Any gap that is >= will require retransmission to guarantee at most "gap" consec packets are lost. This was not my choice, but this is how it currently works. Hence my comment that QoS cube selection will need an overhaul eventually to select qos cubes either by name or by a sound comparison of the qosspec given by the app. Note: eventually. This was a fix to select unreliable using the current (less-than-ideal) consensus.

**Pull request merged**